6PE: nil pointer dereference upon "Peer Up"

Hi guys,

Today I updated gobgp, in order to use this new feature `https://github.com/osrg/gobgp/issues/1086`

Unfortunately, when the session gets established a runtime error occurs. More specifically:
```
{"level":"info","msg":"gobgpd started","time":"2016-09-27T10:38:34+02:00"}
{"Topic":"Config","level":"info","msg":"Finished reading the config file","time":"2016-09-27T10:38:34+02:00"}
{"level":"info","msg":"Peer x.x.x.x is added","time":"2016-09-27T10:38:34+02:00"}
{"Topic":"Peer","level":"info","msg":"Add a peer configuration for:x.x.x.x","time":"2016-09-27T10:38:34+02:00"}

{"Key":"x.x.x.x","State":"BGP_FSM_OPENCONFIRM","Topic":"Peer","level":"info","msg":"Peer Up","time":"2016-09-27T10:38:53+02:00"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x74aa62]

goroutine 22 [running]:
panic(0xd6dd80, 0xc820012140)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/osrg/gobgp/table.CidrToRadixkey(0xc8201ebe40, 0x11, 0x0, 0x0)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/table/destination.go:62 +0xb2
github.com/osrg/gobgp/table.NewDestination(0x7ff40d5915e0, 0xc82023ac80, 0x0, 0x0, 0x0, 0x0)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/table/destination.go:138 +0x259
github.com/osrg/gobgp/table.(*Table).getOrCreateDest(0xc8202a4d10, 0x7ff40d5915e0, 0xc82023ac80, 0x4111e7)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/table/table.go:220 +0x2b4
github.com/osrg/gobgp/table.(*Table).insert(0xc8202a4d10, 0xc8202cacc0, 0x20004)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/table/table.go:72 +0x69
github.com/osrg/gobgp/table.(*TableManager).ProcessPaths(0xc8201ed080, 0xc820132cd8, 0x1, 0x1, 0xc8201243c0, 0x1, 0x1, 0x0, 0x0, 0x0, ...)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/table/table_manager.go:272 +0x252
github.com/osrg/gobgp/server.(*BgpServer).propagateUpdate(0xc8201ac400, 0xc8201b9720, 0xc8201243c0, 0x1, 0x1, 0x0, 0x0, 0x0)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/server/server.go:571 +0x16ec
github.com/osrg/gobgp/server.(*BgpServer).handleFSMMessage(0xc8201ac400, 0xc8201b9720, 0xc820222980)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/server/server.go:749 +0x1dbb
github.com/osrg/gobgp/server.(*BgpServer).Serve.func1(0xc820222980)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/server/server.go:162 +0x49d
github.com/osrg/gobgp/server.(*BgpServer).Serve(0xc8201ac400)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/server/server.go:246 +0x442
created by main.main
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/gobgpd/main.go:180 +0xbcd
```

It occurs every time that am establishing the session. Could you please advise ?

Thanks @ishidawataru for looking into this.

```
$ sudo -E bin/gobgpd -f ../cfg/router.cfg  —-pprof-disable --api-hosts=0.0.0.0:50051 -l=debug
{"level":"info","msg":"gobgpd started","time":"2016-09-27T11:01:06+02:00"}
{"Topic":"Config","level":"info","msg":"Finished reading the config file","time":"2016-09-27T11:01:06+02:00"}
{"level":"info","msg":"Peer a.a.a.a is added","time":"2016-09-27T11:01:06+02:00"}
{"Topic":"Peer","level":"info","msg":"Add a peer configuration for:a.a.a.a","time":"2016-09-27T11:01:06+02:00"}
{"Duration":0,"Key":"a.a.a.a","Topic":"Peer","level":"debug","msg":"IdleHoldTimer expired","time":"2016-09-27T11:01:06+02:00"}
{"Key":"a.a.a.a","Topic":"Peer","level":"debug","msg":"state changed","new":"BGP_FSM_ACTIVE","old":"BGP_FSM_IDLE","reason":"idle-hold-timer-expired","time":"2016-09-27T11:01:06+02:00"}

{"Key":"a.a.a.a","Topic":"Peer","level":"debug","msg":"state changed","new":"BGP_FSM_OPENSENT","old":"BGP_FSM_ACTIVE","reason":"new-connection","time":"2016-09-27T11:01:21+02:00"}
{"Key":"a.a.a.a","Topic":"Peer","level":"debug","msg":"state changed","new":"BGP_FSM_OPENCONFIRM","old":"BGP_FSM_OPENSENT","reason":"open-msg-received","time":"2016-09-27T11:01:21+02:00"}
{"Key":"a.a.a.a","State":"BGP_FSM_OPENCONFIRM","Topic":"Peer","level":"info","msg":"Peer Up","time":"2016-09-27T11:01:21+02:00"}
{"Key":"a.a.a.a","Topic":"Peer","level":"debug","msg":"state changed","new":"BGP_FSM_ESTABLISHED","old":"BGP_FSM_OPENCONFIRM","reason":"open-msg-negotiated","time":"2016-09-27T11:01:21+02:00"}
{"Key":"a.a.a.a","Topic":"Peer","attributes":[{"type":1,"value":0},{"type":2,"as_paths":[{"segment_type":2,"num":2,"asns":[2914,9293]}]},{"type":4,"metric":18289},{"type":5,"value":250},{"type":7,"as":9293,"address":"110.79.255.254"},{"type":8,"communities":[190972314,190973313,190974307,190975304,447626880,447627114,447646186]},{"type":9,"value":"x.x.128.96"},{"type":10,"value":["a.a.a.a","y.y.128.18"]},{"type":14,"nexthop":"x.x.128.96","afi":2,"safi":4,"value":[{"prefix":"2001:2e0::/56"}]}],"level":"debug","msg":"received update","nlri":[],"time":"2016-09-27T11:01:21+02:00","withdrawals":[]}
{"Key":"2001:2e0::/32","Topic":"Table","level":"debug","msg":"create Destination","time":"2016-09-27T11:01:21+02:00"}
{"Key":"2001:2e0::/32","Topic":"table","level":"debug","msg":"Processing destination","time":"2016-09-27T11:01:21+02:00"}
{"Topic":"Table","level":"debug","msg":"computeKnownBestPath known pathlist: 1","time":"2016-09-27T11:01:21+02:00"}
{"Data":{"nlri":{"prefix":"2001:2e0::/56"},"attrs":[{"type":1,"value":0},{"type":2,"as_paths":[{"segment_type":2,"num":2,"asns":[2914,9293]}]},{"type":4,"metric":18289},{"type":5,"value":250},{"type":7,"as":9293,"address":"110.79.255.254"},{"type":8,"communities":[190972314,190973313,190974307,190975304,447626880,447627114,447646186]},{"type":9,"value":"x.x.128.96"},{"type":10,"value":["a.a.a.a","y.y.128.18"]},{"type":14,"nexthop":"x.x.128.96","afi":2,"safi":4,"value":[{"prefix":"2001:2e0::/56"}]}],"age":1474966881,"source-id":"a.a.a.a","neighbor-ip":"a.a.a.a"},"Key":"a.a.a.a","Topic":"Peer","level":"debug","msg":"From same AS, ignore.","time":"2016-09-27T11:01:21+02:00"}
{"Key":"a.a.a.a","Topic":"Peer","attributes":[{"type":1,"value":0},{"type":2,"as_paths":[{"segment_type":2,"num":2,"asns":[6939,23911]}]},{"type":5,"value":250},{"type":8,"communities":[447626880,447626992,447644986]},{"type":9,"value":"x.x.128.10"},{"type":10,"value":["a.a.a.a"]},{"type":14,"nexthop":"x.x.128.10","afi":2,"safi":4,"value":[{"prefix":"2001:252::/56"}]}],"level":"debug","msg":"received update","nlri":[],"time":"2016-09-27T11:01:21+02:00","withdrawals":[]}
{"Key":"2001:252::/32","Topic":"Table","level":"debug","msg":"create Destination","time":"2016-09-27T11:01:21+02:00"}
{"Key":"2001:252::/32","Topic":"table","level":"debug","msg":"Processing destination","time":"2016-09-27T11:01:21+02:00"}
{"Topic":"Table","level":"debug","msg":"computeKnownBestPath known pathlist: 1","time":"2016-09-27T11:01:21+02:00"}
{"Data":{"nlri":{"prefix":"2001:252::/56"},"attrs":[{"type":1,"value":0},{"type":2,"as_paths":[{"segment_type":2,"num":2,"asns":[6939,23911]}]},{"type":5,"value":250},{"type":8,"communities":[447626880,447626992,447644986]},{"type":9,"value":"x.x.128.10"},{"type":10,"value":["a.a.a.a"]},{"type":14,"nexthop":"x.x.128.10","afi":2,"safi":4,"value":[{"prefix":"2001:252::/56"}]}],"age":1474966881,"source-id":"a.a.a.a","neighbor-ip":"a.a.a.a"},"Key":"a.a.a.a","Topic":"Peer","level":"debug","msg":"From same AS, ignore.","time":"2016-09-27T11:01:21+02:00"}
{"Key":"a.a.a.a","Topic":"Peer","attributes":[{"type":1,"value":0},{"type":2,"as_paths":[{"segment_type":2,"num":3,"asns":[6939,23911,9406]}]},{"type":5,"value":250},{"type":8,"communities":[447626880,447626992,447644986]},{"type":9,"value":"x.x.128.10"},{"type":10,"value":["a.a.a.a"]},{"type":14,"nexthop":"x.x.128.10","afi":2,"safi":4,"value":[{"prefix":"2001:250:204::/72"}]}],"level":"debug","msg":"received update","nlri":[],"time":"2016-09-27T11:01:21+02:00","withdrawals":[]}
{"Key":"2001:250:204::/48","Topic":"Table","level":"debug","msg":"create Destination","time":"2016-09-27T11:01:21+02:00"}
{"Key":"2001:250:204::/48","Topic":"table","level":"debug","msg":"Processing destination","time":"2016-09-27T11:01:21+02:00"}
{"Topic":"Table","level":"debug","msg":"computeKnownBestPath known pathlist: 1","time":"2016-09-27T11:01:21+02:00"}
{"Data":{"nlri":{"prefix":"2001:250:204::/72"},"attrs":[{"type":1,"value":0},{"type":2,"as_paths":[{"segment_type":2,"num":3,"asns":[6939,23911,9406]}]},{"type":5,"value":250},{"type":8,"communities":[447626880,447626992,447644986]},{"type":9,"value":"x.x.128.10"},{"type":10,"value":["a.a.a.a"]},{"type":14,"nexthop":"x.x.128.10","afi":2,"safi":4,"value":[{"prefix":"2001:250:204::/72"}]}],"age":1474966881,"source-id":"a.a.a.a","neighbor-ip":"a.a.a.a"},"Key":"a.a.a.a","Topic":"Peer","level":"debug","msg":"From same AS, ignore.","time":"2016-09-27T11:01:21+02:00"}
{"Key":"a.a.a.a","Topic":"Peer","attributes":[{"type":1,"value":0},{"type":2,"as_paths":null},{"type":5,"value":500},{"type":8,"communities":[447617711,447645809,4294967041]},{"type":9,"value":"x.x.198.40"},{"type":10,"value":["a.a.a.a","y.y.128.92"]},{"type":14,"nexthop":"x.x.198.40","afi":2,"safi":4,"value":[{"prefix":"81.210.141.48/150"}]}],"level":"debug","msg":"received update","nlri":[],"time":"2016-09-27T11:01:21+02:00","withdrawals":[]}
{"Key":"81.210.141.48/126","Topic":"Table","level":"debug","msg":"create Destination","time":"2016-09-27T11:01:21+02:00"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x74aa62]

goroutine 6 [running]:
panic(0xd6dd80, 0xc820012140)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/osrg/gobgp/table.CidrToRadixkey(0xc820338660, 0x11, 0x0, 0x0)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/table/destination.go:62 +0xb2
github.com/osrg/gobgp/table.NewDestination(0x7f9c987a9e98, 0xc8201dfb40, 0x0, 0x0, 0x0, 0x0)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/table/destination.go:138 +0x259
github.com/osrg/gobgp/table.(*Table).getOrCreateDest(0xc8202990f0, 0x7f9c987a9e98, 0xc8201dfb40, 0x4111e7)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/table/table.go:220 +0x2b4
github.com/osrg/gobgp/table.(*Table).insert(0xc8202990f0, 0xc82006d5c0, 0x20004)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/table/table.go:72 +0x69
github.com/osrg/gobgp/table.(*TableManager).ProcessPaths(0xc8201df040, 0xc820138cd8, 0x1, 0x1, 0xc8201329b0, 0x1, 0x1, 0x0, 0x0, 0x0, ...)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/table/table_manager.go:272 +0x252
github.com/osrg/gobgp/server.(*BgpServer).propagateUpdate(0xc82000d800, 0xc8201a95e0, 0xc8201329b0, 0x1, 0x1, 0x0, 0x0, 0x0)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/server/server.go:571 +0x16ec
github.com/osrg/gobgp/server.(*BgpServer).handleFSMMessage(0xc82000d800, 0xc8201a95e0, 0xc8200a3480)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/server/server.go:749 +0x1dbb
github.com/osrg/gobgp/server.(*BgpServer).Serve.func1(0xc8200a3480)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/server/server.go:162 +0x49d
github.com/osrg/gobgp/server.(*BgpServer).Serve(0xc82000d800)
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/server/server.go:246 +0x442
created by main.main
	/home/app/netops/gobgp/src/github.com/osrg/gobgp/gobgpd/main.go:180 +0xbcd
```

--- a/packet/bgp/bgp.go
+++ b/packet/bgp/bgp.go
@@ -833,6 +833,10 @@ func NewIPAddrPrefix(length uint8, prefix string) *IPAddrPrefix {
 	}
 }
 
+func isIPv4MappedIPv6(ip net.IP) bool {
+	return len(ip) == net.IPv6len && ip.To4() != nil
+}
+
 type IPv6AddrPrefix struct {
 	IPAddrPrefix
 }
@@ -842,18 +846,11 @@ func (r *IPv6AddrPrefix) AFI() uint16 {
 }
 
 func (r *IPv6AddrPrefix) String() string {
-	isZero := func(p net.IP) bool {
-		for i := 0; i < len(p); i++ {
-			if p[i] != 0 {
-				return false
-			}
-		}
-		return true
-	}(r.Prefix[0:10])
-	if isZero && r.Prefix[10] == 0xff && r.Prefix[11] == 0xff {
-		return fmt.Sprintf("::ffff:%s/%d", r.Prefix.String(), r.Length)
+	prefix := r.Prefix.String()
+	if isIPv4MappedIPv6(r.Prefix) {
+		prefix = "::ffff:" + prefix
 	}
-	return fmt.Sprintf("%s/%d", r.Prefix.String(), r.Length)
+	return fmt.Sprintf("%s/%d", prefix, r.Length)
 }
 
 func NewIPv6AddrPrefix(length uint8, prefix string) *IPv6AddrPrefix {
@@ -1379,7 +1376,21 @@ func (l *LabeledIPAddrPrefix) Serialize() ([]byte, error) {
 }
 
 func (l *LabeledIPAddrPrefix) String() string {
-	return fmt.Sprintf("%s/%d", l.Prefix.String(), int(l.Length)-l.Labels.Len()*8)
+	prefix := l.Prefix.String()
+	if isIPv4MappedIPv6(l.Prefix) {
+		prefix = "::ffff:" + prefix
+	}
+	return fmt.Sprintf("%s/%d", prefix, int(l.Length)-l.Labels.Len()*8)
+}
+
+func (l *LabeledIPAddrPrefix) MarshalJSON() ([]byte, error) {
+	return json.Marshal(struct {
+		Prefix string   `json:"prefix"`
+		Labels []uint32 `json:"labels"`
+	}{
+		Prefix: l.String(),
+		Labels: l.Labels.Labels,
+	})
 }
 
 func NewLabeledIPAddrPrefix(length uint8, prefix string, label MPLSLabelStack) *LabeledIPAddrPrefix {

--- a/packet/bgp/bgp_test.go
+++ b/packet/bgp/bgp_test.go
@@ -538,3 +538,43 @@ func Test_CompareFlowSpecNLRI(t *testing.T) {
 	assert.Nil(err)
 	assert.True(r < 0)
 }
+
+func Test_NLRIwithIPv4MappedIPv6prefix(t *testing.T) {
+	assert := assert.New(t)
+	n1 := NewIPv6AddrPrefix(120, "::ffff:10.0.0.1")
+	buf1, err := n1.Serialize()
+	assert.Nil(err)
+	n2, err := NewPrefixFromRouteFamily(RouteFamilyToAfiSafi(RF_IPv6_UC))
+	assert.Nil(err)
+	err = n2.DecodeFromBytes(buf1)
+	assert.Nil(err)
+	buf2, _ := n2.Serialize()
+	if reflect.DeepEqual(n1, n2) {
+		t.Log("OK")
+	} else {
+		t.Error("Something wrong")
+		t.Error(len(buf1), n1, buf1)
+		t.Error(len(buf2), n2, buf2)
+		t.Log(bytes.Equal(buf1, buf2))
+	}
+
+	label := NewMPLSLabelStack(2)
+
+	n3 := NewLabeledIPv6AddrPrefix(120, "::ffff:10.0.0.1", *label)
+	buf1, err = n3.Serialize()
+	assert.Nil(err)
+	n4, err := NewPrefixFromRouteFamily(RouteFamilyToAfiSafi(RF_IPv6_MPLS))
+	assert.Nil(err)
+	err = n4.DecodeFromBytes(buf1)
+	assert.Nil(err)
+	buf2, _ = n3.Serialize()
+	t.Log(n3, n4)
+	if reflect.DeepEqual(n3, n4) {
+		t.Log("OK")
+	} else {
+		t.Error("Something wrong")
+		t.Error(len(buf1), n3, buf1)
+		t.Error(len(buf2), n4, buf2)
+		t.Log(bytes.Equal(buf1, buf2))
+	}
+}

--- a/table/destination_test.go
+++ b/table/destination_test.go
@@ -399,6 +399,7 @@ func TestRadixkey(t *testing.T) {
 	assert.Equal(t, "000010100000001100100000", CidrToRadixkey("10.3.32.0/24"))
 	assert.Equal(t, "000010100000001100100000", IpToRadixkey(net.ParseIP("10.3.32.0").To4(), 24))
 	assert.Equal(t, "000010100000001100100000", IpToRadixkey(net.ParseIP("10.3.32.0").To4(), 24))
+	assert.Equal(t, CidrToRadixkey("::ffff:0.0.0.0/96")+"000010100000001100100000", CidrToRadixkey("::ffff:10.3.32.0/120"))
 }
 
 func TestMultipath(t *testing.T) {

